### DISCUSSION TITLE
Forced 2FA (3FA) is making me consider leaving github

### DISCUSSION THREAD

**Discussion by ftphikari** [community/community#50539] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

2FA already exists, it's called an email. Steam makes you do 2FA by making you enter correct password for your account, and then sends you a verification code to email. That's called a 2 factor authentication method, it verifies that I have access to my Steam account, as well that I have access to my email. If both email and account were compromised - that's on you.
Steam had this system for years. GitHub however completely disregards the fact that I have an email unless to send me some promo info I don't care about 99% of the time.
Guys, I don't have a phone, *to have less attack surface*, and I don't want to add another verification method to my browser, because I actually keep my account secure by *not keeping a lot of data in the browser*. Having more accounts isn't helping me be more secure, having a good password with an email backup that has a good password does.

Anyway, at least allow the option to send codes to email. I'm not going to install another piece of software to get "more secure". If this change will be forced without this option, I will just leave GitHub for sr.ht or shithub for all I care. Maybe I'll even switch to Fossil, always wanted to try it.

**Comment by Obliczeniowo** [comment]
Same for me, having email verification code I get notification that someone try to login to my account and my password was compromised so I can take proper actions for that. And now I get no notification about that. Sad story is that in settings of notification there is no option to set notify me when someone try to login to my account.

Another thing that this **Authy** I have doubt that this one is secure as it is generating offline access code. So it must be generated by App it's self. This leads to situation that someone potentially can hack the things out. The prove of this authentication is going on without internet access is that I have almost always disabled internet connection on my cellphone and I was surprised by fact that I can get this authentication code.

**Reply by sidb999** [reply]
Everyone knows that email **is** the second factor. Let's expect a disaster on October 14. No one knows how many important repositories are owned by users who don't use that "new" dumb social-network web interface, and didn't see that notification. Notification that forcing you to give a phone number to the microsoft, or they will block your **Github** account. No way. I'm leaving.


P.S.: did anyone really expect that microsoft's purchase of the Github could end up being a good thing for users and the community? We should think about moving before it's too late. No one knows what changes await us tomorrow.

**Reply by kuroodo** [reply]
Every service allows users to use email 2FA, that I am aware of. Google doesn't even require 2FA for the Google Play Developer Console. If someone gains access to that account, I would be absolutely f*****. I have not yet had any security issues with my developer account in the past 5 years. Why? Because even if you managed to somehow get my password, one does not simply log into my gmail account with just my email and password. There are verification steps you first have to take, all while each of my emails and devices get blasted with a notification asking if I'm the one trying to log in.

Even Epic Games allows email 2FA despite how prone it is of a target for hacks and attacks. I'm talking accounts used for actual commercial products using Unreal Engine here. Hell, Epic doesn't even make an effort to make user and developer accounts separate.
Someone gaining access to your Epic Games account would mean that the attacker would have access to all of your Unreal marketplace library, and probably any published marketplace assets. If they change your password, you could potentially lose access to thousands+ dollars worth of assets and more.

These companies aren't idiotic companies and trust their users with the ability to leave their accounts secure. It's insane to me that GitHub doesn't allow email 2FA and have an active stance against it.

**Comment by bwachter** [comment]
From the methods offered by Github only SMS is somewhat acceptable - but I've seen issues with SMS when roaming, so I might be losing access to my account when traveling in the future. 

Security keys are not an option - my browsers run isolated from everything else for security reasons, and don't have access to any hardware components that'd be used as security keys. Enabling access to that for github would make my systems overall less secure. 

Mobile phone based apps don't work either - I treat my phone as mostly insecure, and disposable. 

> For security reasons, GitHub Support may not be able to restore access to accounts with 2FA enabled if you lose
your 2FA credentials and lose access to your account recovery methods. 

This section reads like if I have SMS active, and lose access to the phone number I'll also lose access to my account - so pretty much Github forces me into a situation where my carrier can decide if I get to keep the account.

**Comment by airtower-luna** [comment]
To correct (some of) the misinformation in this threat:

* You don't need a phone to use TOTP. Any [RFC 6238](https://datatracker.ietf.org/doc/html/rfc6238) compliant implementation will work (you could even make your own if you want, but there are a lot already). Mozilla recommended [KeePassXC](https://keepassxc.org/) for people who can't or don't want to use a phone app when they started requiring 2FA for Firefox add-on developers. Mind that keeping TOTP secret and password in the same DB will reduce the security benefit somewhat, consider the security/convenience tradeoff for yourself.
* The fact that TOTP works offline is part of why it's more secure than SMS. There's nothing that could be intercepted by an attacker while generating the code. Yes, security depends on keeping the secret you got during initial setup secret. The only other thing you need is (mostly) accurate time.

**Reply by czarmayne** [reply]
Do you have any idea or guide that can help me implement this new change?
I have a GitHub account used mainly for automation. Thus, enabling 2FA will break the automated task.
I am struggling to find good documentation on how to resolve this issue. Thanks!

**Reply by airtower-luna** [reply]
Depends a bit on what kind of automation you do, but if you use the API you'll be able to use [a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) or OAuth as before. Same for repository access with SSH keys. 2FA is only for login.

**Reply by ghost** [reply]
That's great and all, the issue is that Microsoft has decided that Security is the only thing of any importance, and they've wrecked convenience to a level where people like me who would otherwise have continued contributing on this platform are deciding to jump ship to somewhere that doesn't have such a blatant disrespect for our time.

**Comment by osirisgothra** [comment]
This post is a bit dated, I don't come to github often anymore, but happened to be offloading today and saw it:

I think 2fa is less about security to these companies and more about making sure they can track you to a real-world verifiable identity. You. Personally I feel like it is blatant invasion of privacy on so many levels. I don't want my real-life stuff mixed with my online stuff more than it already is. But they would have you believe that anyone who says that is an active illegal porn trader. 

I'm not doing anything illegal: I just don't like the idea of slowly selling away small bits of freedoms. And when it comes to security, I should have the right to choose what kind of security I think is enough for me. 

Of the many "but why" questions you may ask here's an example: It opens more doors (for them) to later tote away other pieces of our rights later on in the future. Small they are individually, but together can end up being something big. People tend not realize that large organizations use many "reasonable" actions as stepping stones to do something that would be _unreasonable_ had they done it all at once. If you can't figure out why that is a problem then you are likely beyond my ability to convince you. That's fine. To each their own, ignorance is bliss, so on and so forth.

This should be something we can opt-out of by signing some sort of IDGAF contract with our profiles. But we all know where this is going, with MS at the helm, it is a sinking ship just like everything else they buy. This among other reasons is why I have slowly left here for the most part and keep most of my code on private servers.

**Reply by darktjm** [reply]
Dated?  I just got the email saying I'll be forced to use 2fa for future use of github a week ago.  Only now have I started to look into a way to appeal this decision.   I guess there isn't.  The only way is to leave github.  Again.  After I just switched back.  I guess I'll just continue updating my existing git repositories, and never create a new one or submit or respond to issues on other repositories ever again.

Security starts not with stronger authentication, but with **code review**.  If a particular project decides to allow direct contributions without review, they can feel free to set their security however they like.  I will not tolerate this 2fa security theater.  Not for my own, personal, essentially disposable account.  Not just so I can post and respond to other projects' issues.

**Comment by same-101** [comment]
me too i will leave github and move to gitkraken or gitlab

**Reply by kuroodo** [reply]
GitLab has built in GitHub -> GitLab migration

**Comment by ghost** [comment]
Run as fast as you can to Gitlab or any other vendor. Github has lost their minds and their staff are obviously garbage and can't think at a higher order than "oh it's policy"

**Comment by jbglaw** [comment]
So I'm forced to use 2FA now as well. So I'll just delete my forked repos and place anything that's worth it on my own host and delete my account. Thanks, bye...
